python/rascaline/rascaline/utils/clebsch_gordan/_utils.py

```diff
@@ -636,3 +478,98 @@ def cg_tensor_product_blocks_same_samples(
         results.append(block)
 
     return results
+
+
+def _match_samples_of_blocks(
+    block_1: TensorBlock, block_2: TensorBlock
+) -> Tuple[TensorBlock, TensorBlock]:
+    """
+    Reshapes the samples axes of either ``block_1`` or ``block_2`` to match the samples
+    of the other.
+
+    Assumes that the samples dimensions of the block with fewer dimensions are a subset
+    of the dimensions of the other. If the dimensions are not a subset, an error is
+    raised.
+
+    TODO: implement for samples dimensions that are not a subset of the other.
+    """
+    # If the number of dimensions are the same, check they are equivalent and return
+    if len(block_1.samples.names) == len(block_2.samples.names):
+        if not block_1.samples == block_2.samples:
+            raise ValueError("Samples dimensions of the two blocks are not equivalent.")
+
+    # First find the block with fewer dimensions. Reorder to have this block on the
+    # 'left' for simplicity, but record the original order for the final output
+    swapped = False
+    if len(block_1.samples.names) > len(block_2.samples.names):
+        block_1, block_2 = block_2, block_1
+        swapped = True
+
+    for name in block_1.samples.names:
+        if name not in block_2.samples.names:
+            raise ValueError(
+                "Samples dimensions of the two blocks are not compatible."
+                " The block with fewer samples dimensions must be a subset"
+                f" of the other block's dimensions. Got: {block_1.samples.names}"
+                f" and {block_2.samples.names}. Offending dimension: {name}"
+            )
+
+    # Broadcast the values of block_1 along the samples dimensions to match those of
+    # block_2
+    dims_2 = [block_2.samples.names.index(name) for name in block_1.samples.names]
+    matches: List[int] = _dispatch.where(
+        _dispatch.all(
+            block_2.samples.values[:, dims_2][:, None] == block_1.samples.values,
+            axis=2,
+        )
+    )[1].tolist()
+
+    # Build new block and return
+    block_1 = TensorBlock(
+        values=block_1.values[matches],
+        samples=block_2.samples,
+        components=block_1.components,
+        properties=block_1.properties,
+    )
+    if swapped:
+        return block_2, block_1
+
+    return block_1, block_2
+
+
+def _compute_labels_full_cartesian_product(
+    labels_1: Labels,
+    labels_2: Labels,
+) -> Labels:
+    """
+    Computes the full cartesian product of two arbitrary :py:class:`Labels` objects.
+
+    In contrast to the :py:func:`_precompute_keys_full_product` function, this function
+    has no awareness of angular channel combination rules or value matching. It simply
+    takes all combinations of the labels entries in ``labels_1`` and ``labels_2``.
+
+    This function assumes that there are no shared labels dimension names between
+    ``labels_1`` and ``labels_2``.
```

Comment:
there should be an assert for this in the code, to catch any bad use of this function early!